R/infrastructure-git.R

```diff
@@ -29,38 +29,90 @@ use_git <- function(message = "Initial commit", pkg = ".") {
+#'   "origin" remote has the form
+#'   \code{https://github.com/<USERNAME>/<REPO>.git} (\code{protocol = "https"})
+#'   or \code{git@@github.com:<USERNAME>/<REPO>.git} (\code{protocol = "ssh"},
+#'   the default). Read GitHub's help
+#'   \href{https://help.github.com/articles/which-remote-url-should-i-use/}{Which
+#'    remote URL should I use?} for more information.
 #'
 #' @inheritParams install_github
 #' @inheritParams use_git
 #' @param private If \code{TRUE}, creates a private repository.
+#' @param protocol transfer protocol, either "ssh" (the default) or "https"
 #' @family git infrastructure
 #' @keywords internal
 #' @export
-use_github <- function(auth_token = github_pat(), private = FALSE, pkg = ".") {
+#' @examples
+#' \dontrun{
+#' create("testpkg")
+#' use_github(pkg = "testpkg", protocol = "https", private = TRUE)
+#' }
+use_github <- function(auth_token = github_pat(), private = FALSE, pkg = ".",
+                       protocol = c("ssh", "https")) {
   if (is.null(auth_token)) {
     stop("GITHUB_PAT required to create new repo")
   }
 
+  protocol <- match.arg(protocol)
+
   pkg <- as.package(pkg)
   use_git(pkg = pkg)
 
-  r <- git2r::repository(pkg$path)
-  if ("origin" %in% git2r::remotes(r))
+  if (uses_github(pkg$path))
     return(invisible())
 
-  message("Creating GitHub repository")
-
+  message("* Creating GitHub repository")
   create <- github_POST("user/repos", pat = auth_token, body = list(
     name = jsonlite::unbox(pkg$package),
-    description = jsonlite::unbox(pkg$description),
+    #description = jsonlite::unbox(gsub("\n", " ", pkg$description)),
+    description = jsonlite::unbox(pkg$title),
     private = jsonlite::unbox(private)
   ))
 
-  message("Adding remote to GitHub")
-  git2r::remote_add(r, "origin", create$ssh_url)
-  # git2r::branch_set_upstream(git2r::head(r), "origin/master")
-  # git2r::push(r, "origin", "refs/heads/master")
+  message("* Adding GitHub remote")
+  r <- git2r::repository(pkg$path)
+  ## use https universally at this point! we deal with ssh case below
+  git2r::remote_add(r, "origin", create$clone_url)
+
+  message("* Adding GitHub links to DESCRIPTION")
+  use_github_links(pkg$path)
+  if (git_uncommitted(pkg$path)) {
+    git2r::add(r, "DESCRIPTION")
+    git2r::commit(r, "Add GitHub links to DESCRIPTION")
+  }
+
+  message("* Pushing to GitHub and setting remote tracking branch")
+  ## in https case, when GITHUB_PAT is passed as password,
+  ## the username is immaterial, but git2r doesn't know that
+  cred <- git2r::cred_env("DUMMY_EMAIL", "PAT")
+  with_envvar(c("DUMMY_EMAIL" = "whatever", "PAT" = auth_token),
+              git2r::push(r, "origin", "refs/heads/master", credentials = cred))
+  ## now change remote URL if user requested ssh
+  if(protocol == "ssh") {
+    git2r::remote_remove(r, "origin") # git2r doesn't expose URL modification :(
+    git2r::remote_add(r, "origin", create$ssh_url)
+  }
+
+  git2r::branch_set_upstream(git2r::head(r), "origin/master")
```

Comment:
Yes, I imagine so. Currently DESCRIPTION say `git2r (>= 0.10.1)` but I am working against `0.11.0` from CRAN and suspect this actually matters.

In fact, you might want `git2r` to have a CRAN release before you merge this. There are at least 2 changes in v0.11.0.9000 that were prompted by issues I opened and that, if exploited, would make this code cleaner. Notes mostly for me:

  * Ability to edit a remote's URL (https://github.com/ropensci/git2r/commit/26276693ab80d7cce4171786f04c467c66f7d079) vs. delete it then add it back
  * New credential method that works with PAT only (https://github.com/ropensci/git2r/commit/a91a716fc6d7bcc7177e7bdbd942ff646f960f72) vs dummy email/username and PAT as password